Uses Keystone v2.0 identity API and allows choosing the preferred Swift region

liamjack opened the pull request:
This solves [Issue #532](https://github.com/syncany/syncany/issues/532), the previous plugin version only supported Keystone v1.0 Identity API (which is now deprecated). This pull request adds the `tenantName` setting which is required for authentication with v2.0 of the identity API.

This pull request also adds the possibility of choosing the preferred Swift region to choose where to store the data geographically.

Commit:
Supports Keystone v2.0 aswell as choice of preferred Swift region

Commit:
Bumped to version 0.4.1-alpha

pimotte commented:
Awesome!

This update would cause an incompatibility with existing repos (as far as I know, some of which function). Is there a way to allow this change without dropping compatibility on the Syncany side? If not, I don't think there is a problem per se, but I just want to be clear on what the effects of merging this are.

Does everyone who uses swift know what a "Tenant Name" is?

liamjack commented:
Maybe allow a choice of `keystoneVersion` during setup, if left blank default to `v2.0`, if not then use specified version, can be done easily with an `if` or two.

No, I had trouble finding out what it was when I started using Swift, I still don't know what it actually means, I just know where to find it and that I need it to authenticate.

liamjack commented:
Allowing a choice of `keystoneVersion` during setup would require some kind of if - else system inside the Settings, for example:

```
if(tenantVersion.equals("v2.0")) {
    // ask user for tenant name
}
```

Is this possible with the current system ?

pimotte commented:
We could have an optional parameter keystoneVersion in the settings, with default value "v2.0". We could then retrieve that parameter during init, and pass on the tenantName if it's there, raise an exception if it's not (if v2.0) and otherwise ignore it. This would mean making tenantName an optional parameter.

However, since we are in alpha, I'm personally more favoured towards not supporting deprecated stuff. 

I want to leave this open for a couple of days to allow @cro the opportunity to chime in. If he has no preference, I'll probably want to merge as-is.

liamjack commented:
I agree on forgetting about v1.0, Keystone is actually on v3.0 for the Identity API, v2.0 is still supported due to the fact that most cloud hosting solutions do not keep up to date.

mlanner commented:
Just for clarification, the OpenStack API, which Keystone partially implements as part of being the auth system service for OpenStack, is currently using v2 and v3. However, while v1 may have been deprecated in Keystone, it's still very much in use in many Swift deployments that don't use Keystone as its auth backend. There's no requirement that Keystone be used as the auth system for Swift. Therefore, it would probably not be wise to remove v1 compatibility. Leaving the choice of what auth version to use to the user during setup is probably best practice. Alternatively, you can "guess" the auth version being used by looking at the auth URL, which will look something like: 
  https://swift.example.com/auth/v1.0 or
  https://swift.example.com/auth/v2.0
As for the tenant, it's not required in v1.
Hope this helps. /// mL
On Nov 30, 2015 04:04, Liam Jack notifications@github.com wrote:I agree on forgetting about v1.0, Keystone is actually on v3.0 for the Identity API, v2.0 is still supported due to the fact that most cloud hosting solutions do not keep up to date.

—Reply to this email directly or view it on GitHub.

Commit:
Takes advantage of Swift's pseudo-folders to better organize files

Before: ![before](https://i.imgur.com/LeFp2NI.png)

After: ![after](http://i.imgur.com/idgkiPZ.png)

pimotte commented:
@mlanner Thanks for the explanation. Does the auth url always adhere to this pattern?

@cuonic Looking good! Normally I would advise you to work in a different branch for different features, but given that there are several issues with the swift plugin and I doubt someone else will pick those up, I'm not complaining.

liamjack commented:
@mlanner It appears that I've misunderstood how this all works and I just fixed the plugin for my Openstack provider, which is not good.

Looking at the documentation available [here](http://joss.javaswift.org/authentication.html) this explains pretty well the different authentication methods:
- Authentication via Swift: username + password
- Authentication via Keystone: username + password + tenantId / tenantName

At the moment, if the Openstack provider does not support authentication via Swift, it creates an exception (See [issue 532](https://github.com/syncany/syncany/issues/532)) that means absolutely nothing to the unsuspecting user.

@pimotte Sorry I'm not used to multi-branch git repos, I'll try and create a new branch for any future changes.

mlanner commented:
@pimotte From the top of my head, I think so, but it might not be strictly required. It would be rather convenient if it did, wouldn't it? :)

@cuonic Yes, that's a good summary on that page. Aside from the auth systems mentioned there, there are others that somewhat closely mimics the TempAuth methods, including SWAuth and SwiftStack Auth. The main difference between those two and TempAuth is that they implement security and a few other operational benefits, which TempAuth completely lacks. Lastly, any other auth systems out there, home grown ones, are fairly likely to follow v1 style semantics. 

Disclaimer: The reason I know this is because I work for SwiftStack. We have hundreds of production Swift clusters out there and most of them use our SwiftStack Auth. The SwiftStack product supports both v1 and v2 style auth, though. 

On Nov 30, 2015 9:13 AM, Pim Otte notifications@github.com wrote:

> @mlanner Thanks for the explanation. Does the auth url always adhere to this pattern?
> 
> @cuonic Looking good! Normally I would advise you to work in a different branch for different features, but given that there are several issues with the swift plugin and I doubt someone else will pick those up, I'm not complaining.
> 
> —
> Reply to this email directly or view it on GitHub.

Commit:
Now supports Swift and Keystone authentication with optional tenantName and / or tenantId

liamjack commented:
@mlanner Well that explains that then, glad we have someone who knows the technology well around, because I sure don't ;)

This latest commit keeps the Swift plugin as it was before (compatible with Swift authentication) but also adds on the possibility of using Keystone authentication by providing a `tenantName` and / or `tenantId`.

pimotte commented:
Do all combinations of `tenantName` and/or `tenantId` behave as expected? (How are they expected to behave?)

liamjack commented:
### Using Tenant ID

![syncany-keystone-2](https://cloud.githubusercontent.com/assets/821228/11482763/f783b4b6-97a4-11e5-82f9-e0ad264b9378.png)

### Using Tenant Name

![syncany-keystone-1](https://camo.githubusercontent.com/a3c892620aafd67caa56b28b5f04c2fc27346831/68747470733a2f2f692e696d6775722e636f6d2f625854693333542e706e67)

### Using Tenant Name and Tenant ID

![syncany-keystone-3](https://cloud.githubusercontent.com/assets/821228/11482764/f7894ec6-97a4-11e5-9911-8e98e41beff8.png)

pimotte commented:
Awesome. Looks good to me. I'll leave it around for a little bit to allow for comments, and after that I'll merge. I'd say it is safe to base other stuff you might possibly do off this branch.

liamjack commented:
Sounds good, I'll see if I can fix the other problems I'm having.

binwiederhier commented:
@cuonic Beautiful.
